### ISSUE TITLE
Windows Store applications incorrectly assumed to be console applications

### ISSUE THREAD

**Issue by lzybkr** [PowerShell/PowerShell#9970]
## Steps to reproduce

Run a Windows program installed in `${env:USERPROFILE}\AppData\Local\Microsoft\WindowsApps`, e.g. `iTunes.exe` or `wt.exe` (the new Windows Terminal) from the interactive command line.

## Expected behavior

The process launches in a new window, a new prompt is displayed and you can run new commands in PowerShell.

## Actual behavior

PowerShell waits until the Windows application exits before allowing new commands to be entered. PowerShell is behaving as though the program was a console application.

It makes perfect sense to wait when PowerShell can't determine if the process is a Windows or console application, so a proper should do a better job processing these reparse points to inspect the real executable.

**Comment by DHowett-MSFT** [comment]
This is probably a Windows issue (in the `CreateProcess` _AppExecutionAlias_ extension) honestly. Powershell doesn't control whether new processes attach to existing or allocate new consoles.

**Comment by lzybkr** [comment]
@DHowett-MSFT - possibly, but I was thinking this PowerShell [code](https://github.com/PowerShell/PowerShell/blob/dd7e45fe2ebb358f0833c1c5be0f8f5eb90d2554/src/System.Management.Automation/engine/NativeCommandProcessor.cs#L935) is perhaps missing something.

**Comment by iSazonov** [comment]
I found `MicrosoftEdge.exe` in my WindowsApps folder and the application runs as expected. I installed iTunes - PowerShell waits.
For both SHGetFileInfo() returns zero - "not exe".

**Comment by lzybkr** [comment]
Just a guess - but does that api work with symbolic links? If not, resolving the links might be the fix.

**Comment by iSazonov** [comment]
Both is symlinks:
```powershell
Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          21.06.2019    14:26                AppleInc.iTunes_nzyj5cx40ttqa
d----          30.11.2018    16:57                Microsoft.MicrosoftEdge_8wekyb3d8bbwe
la---          21.06.2019    14:26              0 AppleInc.Defaults.exe ->
la---          21.06.2019    14:26              0 AppleInc.MDCrashReportTool.exe ->
la---          21.06.2019    14:26              0 iTunes.exe ->
la---          30.11.2018    16:57              0 MicrosoftEdge.exe ->
```

**Comment by DHowett-MSFT** [comment]
They’re not actually symbolic links. They are “app execution alias” release points that store a package family name and app model ID instead of a path. Check one out with `fsutil reparsepoint query <filename>`.

**Comment by iSazonov** [comment]
So we need to compare reparsepoint properties for the objects to find difference in their behavior?

**Comment by SteveL-MSFT** [comment]
Doesn't repro with PS7 P5

**Comment by DHowett-MSFT** [comment]
I'm seeing this repro with PS7 RC1 and 2.

**Comment by DHowett-MSFT** [comment]
It looks like the pull request above added support to the filesystem provider for detecting these, but not to the process launcher for not waiting on them to return.

**Comment by iSazonov** [comment]
Yes, the PR above added a check to FileSystem provider only. We can use the same one for AppX application launch too.

**Comment by fcharlie** [comment]
Solving this problem is not complicated. We can directly parse ReparsePoint, and then obtain the PE subsytem of its target, so that's it.

```c++
//https://github.com/fcharlie/bela/blob/756cfafd83f0198f92638a85895f6bd0a02edb59/src/belawin/reparsepoint.cc#L259
std::optional<std::wstring> RealPathEx(std::wstring_view src,
                                       bela::error_code &ec) {
  FileReparser reparser;
  if (!reparser.FileDeviceLookup(src, ec)) {
    if (ec.code == ERROR_NOT_A_REPARSE_POINT) {
      ec.code = 0;
      return std::make_optional(bela::PathAbsolute(src));
    }
    return std::nullopt;
  }
  switch (reparser.buffer->ReparseTag) {
  case IO_REPARSE_TAG_APPEXECLINK:
    if (AppExecTarget target; DecodeAppLink(reparser.buffer, target)) {
      return std::make_optional(std::move(target.target));
    }
    ec = bela::make_error_code(1, L"BAD: unable decode AppLinkExec");
    return std::nullopt;
  case IO_REPARSE_TAG_SYMLINK:
    if (std::wstring target; DecodeSymbolicLink(reparser.buffer, target)) {
      return std::make_optional(std::move(target));
    }
    ec = bela::make_error_code(1, L"BAD: unable decode SymbolicLink");
    return std::nullopt;
  case IO_REPARSE_TAG_GLOBAL_REPARSE:
    if (std::wstring target; DecodeSymbolicLink(reparser.buffer, target)) {
      return std::make_optional(std::move(target));
    }
    ec = bela::make_error_code(1, L"BAD: unable decode Global SymbolicLink");
    return std::nullopt;
  default:
    break;
  }
  return std::make_optional<std::wstring>(src);
}
```

[bela::RealPath](https://github.com/fcharlie/bela/blob/756cfafd83f0198f92638a85895f6bd0a02edb59/src/belawin/realpath.cc#L6)

**Comment by DHowett** [comment]
FWIW, Terminal is working around this (and a few other issues) by shipping a separate stub executable, wt, that exits immediately after spawning the real one. We're not blocked by this any longer. :smile:

**Comment by daxian-dbw** [comment]
This issue was regressed after reverting the code that retrieves `AppExecLink` targets (#16044) due to undocumented API.
This is a regression from 7.1.

**Comment by iSazonov** [comment]
> We will start a conversation with the Windows App team and see if they can have the API documented so that we can have a better experience.

https://github.com/PowerShell/PowerShell/pull/16295#issuecomment-959815020
> From the last conversation with the Windows Store team about this, they said it was by-design to not document that API, because they might change how the mapping goes behind the scene.

**Comment by iSazonov** [comment]
@DHowett-MSFT Do you know a p/invoke to get an information is a running process GUI or CUI?

**Comment by daxian-dbw** [comment]
FYI, the Windows Store App team agreed to expose a new API to get the information about the application type of a Store app (called as "App Execution Alias"). They will consider using Windows App SDK (aka Reunion) as the shipping vehicle for the new API.

**Comment by iSazonov** [comment]
Good news! But Really we need an universal API to get the information for _any_ exe by path.

**Comment by iSazonov** [comment]
After we create new process we could read the process image and get subsystem
https://github.com/cuteant/CuteAnt.Extensions/blob/ce6a8da0a289a16f52de8b4246041e5f607c7548/test/System.Reflection.Metadata.Tests/TestUtilities/LoaderUtilities.cs#L30

**Comment by iSazonov** [comment]
After we create new process we could read the process image and get subsystem:
https://github.com/dotnet/runtime/blob/852405c649c1f26124974c97199053d6d8f2adab/src/libraries/System.Reflection.Metadata/tests/TestUtilities/LoaderUtilities.cs#L30

**Comment by jazzdelightsme** [comment]
@daxian-dbw Can we un-Close this Issue? I am still hitting problems and receiving complaints from people due to this problem, and I don't see the problem getting smaller.

I think PS could actually work around this, by querying the main EXE from the launched process. I believe @jborean93 referred to this here:

> See https://github.com/PowerShell/PowerShell/pull/16295#discussion_r736028929 for a potential solution for determining the final executable path. It relies on Windows to resolve the link using the “documented” method. The only hiccup is that it relies on the ability to spawn a suspended process which .NET doesn’t expose.

However, I don't understand why the need/desire to launch the process suspended. Just launch it, get the handle, call QueryFullProcessImageNameW and inspect that EXE. If it fails because the process has already exited, then great, no need to wait for it, whether GUI or not.

**Comment by DHowett** [comment]
Unsurprisingly, issues which have not been fixed remain issues.

**Comment by zooba** [comment]
Still an issue, now also reported to Python (which now uses app aliases by default).
